<div style="border:solid green 3px; padding: 20px">
<b>Привет Полина! 👋</b>

Меня зовут Александр Базилевич и я буду проверять твой проект. Предлагаю общаться на «ты» 🙂 Если это не удобно на ты - дай знать, и мы перейдем на "вы". Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных

Вижу выполнение всех заданий и наличие выводов. Все ключевые этапы в работе выполнены, так что в целом справиться с задачей тебе удалось. Есть ещё некоторые вопросы, которые требуют твоего внимания. Комментарии по ним помечены <font color='red'>красным цветом</font> ❌. После их доработки проект будет принят. Осталось совсем немного, буду ждать твою работу на повторное ревью 🙂

Давай работать над проектом в диалоге: если ты <b>что-то меняешь</b> в проекте по моим рекомендациям — <b>пиши об этом</b>. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, <b>не перемещай, не изменяй и не удаляй мои комментарии</b>. Всё это поможет выполнить повторную проверку твоего проекта оперативнее

Ты также можешь реагировать на мои комментарии своими, выделяя их цветами и наиболее понравившимся тебе способом оформления, но явно  отличающимся от моих комментариев. Это нужно, чтобы не создавалась путаница 😉

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅</b>

В случае, если решение на отдельном шаге является полностью правильным 👍 Так будут отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера - замечания и рекомендации ⚠️</b>

В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно)
</div>
    
<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера - на доработку ❌</b>   

В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность найти и исправить ее. На реальной работе твой руководитель будет поступать также, и я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку! 🤓
</div> 

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Да, авторское решение и вправду очень изящно! Давай посмотрим что получилось у тебя!
</div>

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:

def rate(row,param):
    data = row.groupby(param).agg({'debt':'sum'})
    data['count'] = row.groupby(param)['debt'].count()
    data['rate'] = data['debt'] / data['count'] * 100
    return data
data_children = rate(data,'children')
data_children.sort_values(by='rate',ascending=True)

debt  count      rate
children                       
5            0      9  0.000000
0         1063  14091  7.543822
3           27    330  8.181818
1          444   4808  9.234609
2          194   2052  9.454191
4            4     41  9.756098

In [23]:
def rate_pivot(variable):
    data_pivot = data.pivot_table(index=variable, values='debt', aggfunc=['count', 'sum', 'mean'])
    data_pivot.columns = ['all_credits', 'all_debtors','share_debtors']
    return data_pivot
rate_pivot('children')

all_credits  all_debtors  share_debtors
children                                         
0               14091         1063       0.075438
1                4808          444       0.092346
2                2052          194       0.094542
3                 330           27       0.081818
4                  41            4       0.097561
5                   9            0       0.000000

<div class="alert alert-success">
<b>Комментарий 2 ревьюера - все отлично! ✅ </b>

Прекрасно! Ты большой молодец - быстро и качественно все сделала 🤗
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Во-первых хочу сказать что ты большой молодец! Все сделано правильно и аккуратно. Просто хотел тебе написать об этом!
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера - на доработку ❌</b>   

Твой код дает правильный результат. Но наша задача еще и научиться быстрым и эффективным алгоритмам, поэтому:
1. Сводные таблицы - главный инструмент датасаентиста. Он делается двумя способами `.groupby` или `.pivot_table`. Первый - более популярен, но не все можно через него сделать. Их основное отличие в том, что при множественной группировке (по двум и более целевым переменным - например дети и образование...) groupby делает сводную таблицу вниз по строкам. А вот pivot_table делает его вширь, по столбцам. Поэтому эти инструменты разные и например для визуализации hitmap нам  в будущем нужен будет именно pivot_table. Поэтому давай попробуем сделать в виде pivot_table. 
    
1. Давай разберемся с данными. Это проще всего сделать одной строкой с помощью чуть более сложной сводной таблицы. Как ты правильно сделала 'Всего кредитополучателей' - это `count`. 'Всего должников' - это `sum`. А вот долю должников можно подсчитать через `mean` (потому что у нас данные 0 или 1 в поле `dept`). Таким образом в `data.pivot_table` агрегация будет иметь вид `aggfunc=['count', 'sum', 'mean']`.

2. Спасибо за функцию, ведь мы 4 раза выполняем один и тот же алгоритм. Поэтому давай попробуем сделать функцию (назови ее как тебе больше нравится, только не системными названиями) на вход которой принимается одна переменная категория (например, `variable`). В `pivot_table` присвой `index=variable`. То есть мы будем в сводную таблицу передавать разные переменные и она будет строить нам разные сводные таблицы
    
Внутри три строки - первая это твой новый код со сводной таблицей из пункта 2 выше. Вторая - это переименование колонок тоже из твоего нового кода пункта 3, что был выше. Третья - вызов функции `display()` с переменной в скобках того имени, которое ты присваиваешь в первом пункте. Возвращать (то есть писать `return`) в конце функции не надо. Мы передаем в функцию переменную столбца датасета, а она нам рассчитывает и выводит на экран нужные данные

При создании сводных таблиц получаются странные названия заголовков. Они на разных уровнях и в несколько строк иногда. Это называется мультииндексом и это особенность нашей програмной среды. Чтобы в дальнейшем работать с полученной таблицей (не в этом задании, но все же), обращаться к данным через названия столбцов необходимо в этой задаче просто переименовать полученные данные таким кодом data.columns = с значениями ['all_credits', 'all_debtors','share_debtors'] или (на будущее) выполнить такой код data.columns = ['_'.join(col).strip() for col in temp.columns.values]. Более подробно тут https://stackoverflow.com/questions/14507794/pandas-how-to-flatten-a-hierarchical-index-in-columns
    
3. А чтобы запустить твою функцию в следующей пустой ячейке напиши имя своей функции и в скобках передай ей аргумент имени колонки датасета `data` (это будет `'children'` или `'family_status'` и так далее)
    
Попробуй это сделать в новых ячейках чтобы сравнить полученные данные
 
</div>

Вывод: Да, люди без детей с большей вероятностью вернут кредит, чем люди без детей (количество семей с 4 и 5 детьми несущественно для анализа).

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Супер! Все верно с выводами по бездетным и по многодетным!
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data_family = rate(data,'family_status')
data_family.sort_values(by='rate',ascending=True)

debt  count      rate
family_status                               
вдовец / вдова           63    951  6.624606
в разводе                84   1189  7.064760
женат / замужем         927  12261  7.560558
гражданский брак        385   4134  9.313014
Не женат / не замужем   273   2796  9.763948

In [25]:
rate_pivot('family_status')

all_credits  all_debtors  share_debtors
family_status                                                 
Не женат / не замужем         2796          273       0.097639
в разводе                     1189           84       0.070648
вдовец / вдова                 951           63       0.066246
гражданский брак              4134          385       0.093130
женат / замужем              12261          927       0.075606

<div class="alert alert-success">
<b>Комментарий 2 ревьюера - все отлично! ✅ </b>

Да, это именно то, что требуется!
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера - на доработку ❌</b>   

Так, здесь очень хорошо, спасибо за расчеты, но давай вызовем тут твою функцию, о которой было написано выше

</div>

Вывод: Да, лучше всего выплачивают долги вдовцы/вдовы и люди в разводе. Хуже всего - люди без отношений. Вероятно, семейная жизнь вынуждает людей быть более дисциплинированными, а также должна быть корреляция между семейным положением и возрастом, возможно, возраст также оказывает влияние на способность расплачиваться по кредиту.

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Согласен с полученными выводами! Все именно так!
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
data_income = rate(data,'total_income_category')
data_income.sort_values(by='rate',ascending=True)

debt  count      rate
total_income_category                       
D                        21    349  6.017192
B                       354   5014  7.060231
A                         2     25  8.000000
C                      1353  15921  8.498210
E                         2     22  9.090909

In [27]:
rate_pivot('total_income_category')

all_credits  all_debtors  share_debtors
total_income_category                                         
A                               25            2       0.080000
B                             5014          354       0.070602
C                            15921         1353       0.084982
D                              349           21       0.060172
E                               22            2       0.090909

<div class="alert alert-success">
<b>Комментарий 2 ревьюера - все отлично! ✅ </b>

Супер!
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера - на доработку ❌</b>   

И здесь все прекрасно! Просто давай добавим вызов твоей обновленной функции

</div>

Вывод: Да, лучше всего расплачиваются по кредитам заемщики категории D. 

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Отлично! Хотя видно что в разных категориях ситуация разная. Тут важно увидеть, что в категории наиболее состоятельных выводы делать некорректно, так как данных очень мало.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
data_purpose = rate(data,'purpose_category')
data_purpose.sort_values(by='rate',ascending=True)

debt  count      rate
purpose_category                               
операции с недвижимостью   780  10751  7.255139
проведение свадьбы         183   2313  7.911803
получение образования      369   3988  9.252758
операции с автомобилем     400   4279  9.347978

In [29]:
rate_pivot('purpose_category')

all_credits  all_debtors  share_debtors
purpose_category                                                 
операции с автомобилем           4279          400       0.093480
операции с недвижимостью        10751          780       0.072551
получение образования            3988          369       0.092528
проведение свадьбы               2313          183       0.079118

<div class="alert alert-success">
<b>Комментарий 2 ревьюера - все отлично! ✅ </b>

Замечательно! Спасибо
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера - на доработку ❌</b>   

И тут все хорошо с полученными данными. Давай просто вызовем новую функцию и сравним, что получилось

</div>

Вывод: С большей вероятностью человек расплатится, если брал кредит для операций с недвижимостью или для свадьбы, с меньшей вероятностью - если кредит был для целей образования или для покупки автомобиля.

<div class="alert alert-warning">
<b>Комментарий ревьюера - замечания и рекомендации ⚠️</b>

Хорошо! Но давай посмотрим в полученную таблицу - там же видно, что у нас две большие группы - недвижимость/свадьба и образование/автокредиты. Давай их добавим в промежуточный вывод!
</div>

<div class="alert alert-success">
<b>Комментарий 2 ревьюера - все отлично! ✅ </b>

Спасибо за уточнение!
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Ответ: Могут быть разные причины:
1) Люди не захотели указывать информацию о себе.
2) Есть группы людей без опыта работы.
3) Техническая ошибка при выгрузке/обработке данных.
4) Человеческий фактор при выгрузке/обработке данных.

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Прекрасный ответ, спасибо! Подвох был в том, что мы работаем с статистикой выданных кредитов. А люди могут умышленно не давать информацию, предполагая, что это создает выйгрышные условия для получения займа
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Ответ: Медианное значение не искажает данные для статистического анализа, в отличие от среднеарифмитического значения, которое может завышать/занижать статистику из-за небольшого количества аномальных значений. 

<div class="alert alert-success">
<b>Комментарий ревьюера - все отлично! ✅ </b>

Да, мы обычно говорим так - вбросы сильно влияют на средне-арифмитическое и существенно меньше на медиану. А на будущее -  надо понимать откуда вбросы получились и обосновано принимать решение по их обработке
</div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.
 
В ходе исследования мы смогли проанализивать кредитоспособность заемщиков в зависимости от различных критериев, для этих целей мы: 
1) Проанализивароли общую информацию о данных.  
2) Избавились от 2174 пропущенных значений к колонках days_employed и total_income, заменив их медианными значениями для категории.
3) Избавились от отрицательных и аномальных значений в столбцах days_employed и children, т.
4) Обработали дубликаты и типы данных в таблице.
5) Распределили всех заемщиков по категориям, в зависимости от размера дохода и цели кредита.

Далее мы выяснили, что на вероятность возврата кредита оказывают влияние различные причины:
1) Количество детей (наибольшей кредитоспособностью обладают заемщики без детей)
2) Семейное положение (наибольшей кредитоспособностью обладают вдовы/вдовцы и семейные люди)
3) Уровень дохода (наибольшей кредитоспособностью обладают заемщики с уровнем дохода от 30 до 50 тр.)
4) Цели кредита (чаще всего заемщики возвращают кредиты, полученные для операций с недвижимостью или для свадьбы)

<div class="alert alert-success">
<b>Комментарий 2 ревьюера - все отлично! ✅ </b>

Великолепно! Никаких вопросов
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера - на доработку ❌</b>   

Прошу в финальный вывод добавить описание данных. Какие были получены данные, сколько их было, с какими ошибками/пропусками мы столкнулись, как обрабатывались. Поля описывать не надо - достаточно показать насколько можно доверять им!
    
Потому что, по моему опыту работу, в первую очередь заказчик исследования читает вывод и только потом изучает весь проект, если есть на то время. А рекомендации это сжатые промежуточные выводы по каждому из разделов. Лучше их дать списком
    
</div>

<div class="alert alert-info">

<h3> Резюме по проекту ❌ <a class="tocSkip"> </h3>
  
Я вижу очень хорошо выполненную работу с наличием выводов! Давай добавим описание данных в финальный вывод и сделаем новую функцию для более простого вывода результатов
  
Остались сделать небольшие доработки, чтобы принять проект:
* сделать новую сводную таблицу в одной функции
* 4 раза ее вызвать  
* поправить промежуточные выводы
* добавить описание данных в финальный вывод

Спасибо! Так держать 👍 Жду твой проект на повторное ревью!
</div>

<div class="alert alert-info">

<h3> Резюме 2 по проекту ✅<a class="tocSkip"> </h3>
  
Полина, спасибо что сделала функцию с сводной таблицей и учла все мои рекомендации. Так держать! 

Работа принята. Удачи с следующими проектами 🍀
</div>